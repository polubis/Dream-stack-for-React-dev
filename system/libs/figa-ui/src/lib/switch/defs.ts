import type { ReactNode, HTMLAttributes, DetailedHTMLProps } from 'react';

type SwitchHTMLElementProps = Omit<
  DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>,
  'children' | 'onClick'
>;

type SwitchProps  = SwitchHTMLElementProps & {
  label?: ReactNode;
  reversed?: boolean;
  active?: boolean;
  disabled?: boolean;
  onClick?(): void;
}

export type { SwitchProps, SwitchHTMLElementProps };
