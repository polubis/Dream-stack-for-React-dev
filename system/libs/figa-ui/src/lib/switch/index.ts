export * from './switch';
export * from './defs';
