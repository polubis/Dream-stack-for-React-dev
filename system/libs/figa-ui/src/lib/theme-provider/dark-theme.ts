import type { Theme } from './defs';
import { tokens } from './tokens';

const dark: Theme = {
  outline: {
    color: tokens.common.white,
  },
  switch: {
    bg: tokens.gray[50],
    color: tokens.common.white,
    active: {
      bg: tokens.green[200],
    },
  },
  scroll: {
    thumb: tokens.gray[600],
    track: tokens.gray[550],
    thumbHover: tokens.gray[650],
  },
  divider: {
    default: {
      bg: tokens.common.white,
    },
    primary: {
      bg: tokens.primary[100],
    },
  },
  radio: {
    borderColor: tokens.common.white,
    active: {
      bg: tokens.green[200],
      borderColor: tokens.green[200],
    },
  },
  checkbox: {
    borderColor: tokens.common.white,
    checked: {
      bg: tokens.primary[50],
      borderColor: tokens.primary[50],
      color: tokens.common.white,
    },
  },
  tabs: {
    filled: {
      bg: tokens.gray[500],
      active: {
        borderColor: tokens.green[50],
      },
    },
  },
  chips: {
    color: tokens.common.white,
    borderColor: tokens.common.white,
    active: {
      color: tokens.green[50],
      borderColor: tokens.green[50],
    },
  },
  badge: {
    outlined: {
      primary: {
        color: tokens.primary[50],
        borderColor: tokens.primary[50],
      },
      secondary: {
        color: tokens.primary[100],
        borderColor: tokens.primary[100],
      },
      ok: {
        color: tokens.green[100],
        borderColor: tokens.green[100],
      },
      casual: {
        color: tokens.gray[150],
        borderColor: tokens.gray[150],
      },
    },
    filled: {
      primary: {
        color: tokens.common.black,
        bg: tokens.primary[50],
      },
      secondary: {
        color: tokens.common.black,
        bg: tokens.primary[100],
      },
      ok: {
        color: tokens.common.black,
        bg: tokens.green[100],
      },
      casual: {
        color: tokens.common.black,
        bg: tokens.gray[150],
      },
    },
  },
  thumbnail: {
    bg: tokens.backdrop[100],
    color: tokens.common.white,
  },
  filePicker: {
    outline: {
      color: tokens.primary[50],
    },
    invalid: {
      color: tokens.red[150],
      outline: tokens.red[150],
    },
  },
  field: {
    invalid: {
      color: tokens.red[100],
    },
    hint: {
      color: tokens.common.white,
    },
  },
  font: {
    default: {
      color: tokens.common.white,
    },
    primary: {
      color: tokens.primary[100],
    },
  },
  body: {
    bg: tokens.common.black,
  },
  alert: {
    filled: {
      info: {
        bg: tokens.gray[150],
        color: tokens.common.black,
      },
      ok: {
        bg: tokens.green[150],
        color: tokens.common.black,
      },
      error: {
        bg: tokens.red[50],
        color: tokens.common.white,
      },
      warn: {
        bg: tokens.yellow[50],
        color: tokens.common.black,
      },
    },
    outlined: {
      info: {
        color: tokens.common.white,
      },
      ok: {
        color: tokens.green[150],
      },
      error: {
        color: tokens.red[150],
      },
      warn: {
        color: tokens.yellow[50],
      },
    },
  },
  modal: {
    bg: tokens.common.black,
    border: tokens.dark[50],
    backdrop: tokens.backdrop[50],
  },
  button: {
    filled: {
      primary: {
        color: tokens.common.black,
        bg: tokens.primary[50],
        outlineColor: tokens.common.white,
      },
      secondary: {
        color: tokens.common.black,
        bg: tokens.primary[100],
        outlineColor: tokens.common.white,
      },
      tertiary: {
        color: tokens.common.black,
        bg: tokens.common.white,
        outlineColor: tokens.common.white,
      },
    },
    outlined: {
      primary: {
        color: tokens.primary[50],
        bg: 'transparent',
        borderColor: tokens.primary[50],
        outlineColor: tokens.common.white,
      },
      secondary: {
        color: tokens.primary[100],
        bg: 'transparent',
        borderColor: tokens.primary[100],
        outlineColor: tokens.common.white,
      },
      tertiary: {
        color: tokens.common.white,
        bg: 'transparent',
        borderColor: tokens.common.white,
        outlineColor: tokens.common.white,
      },
    },
    ghost: {
      hoverBg: tokens.dark[250],
    },
  },
  select: {
    bg: tokens.dark[50],
    color: tokens.common.white,
    placeholderColor: tokens.gray[50],
    hoverBg: tokens.gray[100],
    optionSeparator: tokens.gray[100],
    optionActiveColor: tokens.secondary[50],
  },
  nav: {
    bg: tokens.dark[100],
    borderColor: tokens.gray[300],
    bgWithOpacity: tokens.dark[200],
  },
  logo: {
    text: {
      first: tokens.purple[100],
      second: tokens.common.white,
      last: tokens.secondary[200],
    },
    bg: {
      first: {
        from: tokens.red[50],
        to: tokens.primary[250],
      },
      second: {
        from: tokens.secondary[250],
        to: tokens.primary[250],
      },
      last: {
        from: tokens.blue[200],
        to: tokens.primary[250],
      },
    },
  },
  progressCircle: {
    bg: tokens.dark[50],
  },
  input: {
    filled: {
      bg: tokens.dark[50],
      color: tokens.common.white,
      placeholderColor: tokens.gray[50],
      fxColor: tokens.common.white,
    },
    outlined: {
      borderColor: tokens.common.white,
      color: tokens.common.white,
      placeholderColor: tokens.gray[50],
      fxColor: tokens.common.white,
    },
    empty: {
      color: tokens.common.white,
      placeholderColor: tokens.gray[50],
      fxColor: tokens.common.white,
    },
    invalid: tokens.red[150],
  },
  placeholder: {
    filled: {
      bg: tokens.gray[300],
    },
    outlined: {
      bg: tokens.common.black,
      borderColor: tokens.gray[300],
    },
  },
  box: {
    filled: {
      bg: tokens.gray[300],
    },
    outlined: {
      bg: tokens.common.black,
      borderColor: tokens.gray[300],
    },
  },
  footer: {
    bg: tokens.common.black,
    borderColor: tokens.gray[300],
  },
  code: {
    bg: tokens.dark[150],
    gutters: tokens.dark[150],
    selectionBg: '#4d4d4d',
    foldPlaceholder: '#4d4d4d',
    tags: [
      { tag: 'angleBracket', color: '#da6bad' },
      { tag: 'arithmeticOperator', color: '#fff' },
      { tag: 'bitwiseOperator', color: '#fff' },
      { tag: 'blockComment', color: '#6A9955' },
      { tag: 'bool', color: '#569cd6' },
      { tag: 'brace', color: '#ffd70f' },
      { tag: 'bracket', color: '#da6bad' },
      { tag: 'className', color: '#3abfb0' },
      { tag: 'comment', color: '#6A9955' },
      { tag: 'compareOperator', color: '#fff' },
      { tag: 'controlKeyword', color: '#eba0d9' },
      { tag: 'definition', color: '#4fc1ff' },
      { tag: 'definitionKeyword', color: '#3981c5' },
      { tag: 'definitionOperator', color: '#fff' },
      { tag: 'derefOperator', color: '#fff' },
      { tag: 'function', color: '#dcdc9f' },
      { tag: 'keyword', color: '#3981c5' },
      { tag: 'labelName', color: '#4fc1ff' },
      { tag: 'lineComment', color: '#6A9955' },
      { tag: 'literal', color: '#ef8c59' },
      { tag: 'logicOperator', color: '#fff' },
      { tag: 'moduleKeyword', color: '#eba0d9' },
      { tag: 'name', color: '#4fc1ff' },
      { tag: 'null', color: '#569cd6' },
      { tag: 'number', color: '#accea8' },
      { tag: 'operatorKeyword', color: '#3abfb0' },
      { tag: 'paren', color: '#da6bad' },
      { tag: 'typeName', color: '#3abfb0' },
    ],
  },
  codeBlock: {
    header: {
      bg: tokens.gray[350],
      dots: {
        first: {
          bg: tokens.green[50],
        },
        second: {
          bg: tokens.yellow[50],
        },
        third: {
          bg: tokens.blue[650],
        },
      },
    },
    content: {
      bg: tokens.dark[150],
    },
  },
  list: {
    marker: {
      bg: tokens.primary[50],
    },
  },
  blockquote: {
    bg: tokens.gray[450],
    borderColor: tokens.gray[300],
  },
  loader: {
    primary: tokens.common.white,
    secondary: tokens.primary[50],
  },
  avatars: {
    restCounter: {
      bg: tokens.primary[50],
      color: tokens.common.black,
    },
  },
  creatorLayout: {
    borderColor: tokens.gray[300],
    bg: tokens.common.black,
  },
};

export { dark };
