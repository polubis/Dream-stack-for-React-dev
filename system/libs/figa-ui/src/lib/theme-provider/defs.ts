import type { ReactNode } from 'react';

interface ThemeProviderProps {
  children: ReactNode;
}

interface Theme {
  font: {
    color: string;
  };
  background: {
    body: string;
  };
}

interface Tokens {
  [key: string]: {
    [key: string]: string;
  };
}

type ThemeKey = 'dark' | 'light';
type Themes = Record<ThemeKey, Theme>;
type ThemesList = [ThemeKey, Theme][];

interface ThemeProviderValue {
  key: ThemeKey;
  theme: Theme;
  themes: Themes;
  themesList: ThemesList;
  setTheme: (key: ThemeKey) => void;
}

type UnsafeThemeProviderValue = ThemeProviderValue | null;

type GetDefaultValue = (
  value?: Partial<Pick<ThemeProviderValue, 'key' | 'setTheme'>>
) => ThemeProviderValue;

export type {
  ThemeProviderProps,
  Theme,
  ThemeProviderValue,
  UnsafeThemeProviderValue,
  ThemeKey,
  Tokens,
  Themes,
  ThemesList,
  GetDefaultValue,
};
