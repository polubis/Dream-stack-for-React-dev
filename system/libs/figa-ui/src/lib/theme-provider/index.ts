export * from './theme-provider';
export * from './defs';
