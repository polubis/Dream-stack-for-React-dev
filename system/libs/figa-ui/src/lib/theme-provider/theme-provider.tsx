import { useState, createContext, useMemo, useContext } from 'react';
import { ThemeProvider as StyledThemeProvider } from 'styled-components';
import type {
  ThemeKey,
  ThemeProviderProps,
  ThemeProviderValue,
  ThemesList,
  UnsafeThemeProviderValue,
  GetDefaultValue,
} from './defs';
import { themes } from './themes';
import { GlobalStyle } from './global-style';

const Ctx = createContext<UnsafeThemeProviderValue>(null);

const defaultKey = 'dark' as ThemeKey;

const getDefaultValue: GetDefaultValue = ({
  key = defaultKey,
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  setTheme = () => {},
} = {}) => ({
  key,
  setTheme,
  themes,
  theme: themes[key],
  themesList: Object.entries(themes) as ThemesList,
});

const ThemeProvider = ({ children }: ThemeProviderProps) => {
  const [key, setKey] = useState(defaultKey);

  const value: ThemeProviderValue = useMemo(
    () =>
      getDefaultValue({
        key,
        setTheme: setKey,
      }),
    [key]
  );

  return (
    <Ctx.Provider value={value}>
      <StyledThemeProvider theme={value.theme}>
        <GlobalStyle />
        {children}
      </StyledThemeProvider>
    </Ctx.Provider>
  );
};

const useThemeProvider = (): ThemeProviderValue => {
  const ctx = useContext(Ctx);

  return ctx ? ctx : getDefaultValue();
};

export { ThemeProvider, useThemeProvider };
