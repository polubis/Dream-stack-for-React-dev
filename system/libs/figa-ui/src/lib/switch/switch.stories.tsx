import type { Story, Meta } from '@storybook/react';

import { Switch } from './switch';
import { Box } from '../box';
import { Font } from '../font';

export default {
  component: Switch,
  title: 'Switch',
} as Meta;

const Template: Story = () => {
  return (
    <Box
      padding={[300, 300, 300, 300]}
      spacing={[150, 300, 150, 300, 150, 300, 150, 300]}
    >
      <Font variant="h5">Empty</Font>
      <Switch />
      <Font variant="h5">With labels</Font>
      <Box orientation="row" spacing={[150]}>
        <Switch label="Default" />
        <Switch reversed label="Reversed" />
      </Box>
      <Font variant="h5">Active</Font>
      <Box orientation="row">
        <Switch label="Default" active />
      </Box>
      <Font variant="h5">Disabled</Font>
      <Box orientation="row" spacing={[150]}>
        <Switch label="Default" disabled />
        <Switch label="Default" disabled active />
      </Box>
    </Box>
  );
};

export const Default = Template.bind({});
Default.args = {};
