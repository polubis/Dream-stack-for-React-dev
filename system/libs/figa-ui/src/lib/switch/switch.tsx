import styled from 'styled-components';
import type { SwitchProps } from './defs';
import c from 'classnames';
import { useId } from 'react';
import { center, row, size } from '../shared';
import { Font } from '../font';
import { tokens } from '../theme-provider';

const Container = styled.div`
  ${row()}

  &.reversed {
    flex-direction: row-reverse;

    & > *:first-child {
      margin-left: ${tokens.spacing[150]};
    }
  }

  &:not(.reversed) {
    & > *:first-child {
      margin-right: ${tokens.spacing[150]};
    }
  }

  &.disabled {
    opacity: 0.5;
    cursor: not-allowed;

    .switch-container {
      .switch-button,
      .switch-overlay {
        cursor: not-allowed;
      }
    }
  }

  &:not(.disabled) {
    .switch-container {
      .switch-overlay:active .switch-button {
        width: 60%;
      }
    }
  }

  .switch-container {
    input {
      position: absolute;
      ${size('0px')}
      visibility: hidden;
    }

    .switch-overlay {
      ${center()}
      width: ${tokens.spacing[700]};
      height: ${tokens.spacing[350]};
      cursor: pointer;
      position: relative;
      border-radius: ${tokens.spacing[1000]};
      transition: background-color 0.2s;
      background: ${(props) => props.theme.switch.bg};
    }

    .switch-button {
      content: '';
      ${size(tokens.spacing[250])}
      left: ${tokens.spacing[50]};
      top: 0;
      cursor: pointer;
      bottom: 0;
      margin: auto 0;
      border: none;
      position: absolute;
      border-radius: 45px;
      box-shadow: ${tokens.shadow[50]};
      transition: 0.2s;
      background: ${(props) => props.theme.switch.color};
    }

    .switch-checkbox:checked + .switch-overlay .switch-button {
      left: calc(100% - ${tokens.spacing[50]});
      transform: translateX(-100%);
    }

    .switch-checkbox:checked + .switch-overlay {
      background: ${(props) => props.theme.switch.active.bg};
    }
  }
`;

const Switch = ({
  className,
  label,
  reversed,
  active,
  disabled,
  ...props
}: SwitchProps) => {
  const id = useId();
  // @TODO Find solution for wrong type assignment here later.
  return (
    <Container
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      {...(props as any)}
      className={c('switch', { reversed }, { disabled }, className)}
    >
      <div className="switch-container">
        <input
          id={id}
          type="checkbox"
          disabled={disabled}
          className="switch-checkbox"
          checked={active}
        />
        <label className="switch-overlay" htmlFor={id}>
          <button className="switch-button" />
        </label>
      </div>
      {label && <Font variant="b2">{label}</Font>}
    </Container>
  );
};

export { Switch };
