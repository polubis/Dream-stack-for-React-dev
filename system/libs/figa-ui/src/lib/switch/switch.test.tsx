import { fireEvent, render, screen } from '@testing-library/react';
import { Switch } from './switch';
import { ThemeProvider } from '../theme-provider';

describe('Switch can be used when: ', () => {
  it('[FRAGILE] assigns classes', () => {
    const { asFragment } = render(
      <ThemeProvider>
        <Switch className="my-class" active disabled label="Label" reversed />
      </ThemeProvider>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('allows to use other properties', () => {
    const onClickSpy = jest.fn();
    const onHoverSpy = jest.fn();

    render(
      <ThemeProvider>
        <Switch onClick={onClickSpy} onMouseOver={onHoverSpy} label="Switch" />
      </ThemeProvider>
    );

    fireEvent.click(screen.getByText(/Switch/));
    fireEvent.mouseOver(screen.getByText(/Switch/));

    expect(onClickSpy).toHaveBeenCalledTimes(1);
    expect(onHoverSpy).toHaveBeenCalledTimes(1);
  });
});
