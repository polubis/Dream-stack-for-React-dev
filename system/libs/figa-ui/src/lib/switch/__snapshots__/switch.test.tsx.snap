// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch can be used when:  [FRAGILE] assigns classes 1`] = `
<DocumentFragment>
  <div
    class="sc-bcXHqe cTbyJi switch reversed disabled my-class"
  >
    <div
      class="switch-container"
    >
      <input
        checked=""
        class="switch-checkbox"
        disabled=""
        id=":r0:"
        type="checkbox"
      />
      <label
        class="switch-overlay"
        for=":r0:"
      >
        <button
          class="switch-button"
        />
      </label>
    </div>
    <p
      class="font b2 default"
    >
      Label
    </p>
  </div>
</DocumentFragment>
`;
